File: Flutter/android/src/main/java/com/example/super_player/SuperPlatformPlayerView.java
package com.example.super_player;

import android.content.Context;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;

import com.tencent.liteav.demo.superplayer.SuperPlayerGlobalConfig;
import com.tencent.liteav.demo.superplayer.SuperPlayerModel;
import com.tencent.liteav.demo.superplayer.SuperPlayerVideoId;
import com.tencent.liteav.demo.superplayer.SuperPlayerView;
import com.tencent.liteav.demo.superplayer.model.ISuperPlayerListener;
import com.tencent.rtmp.TXLiveBase;
import com.tencent.rtmp.TXVodPlayer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.platform.PlatformView;

import android.util.Log;

public class SuperPlatformPlayerView implements PlatformView, MethodChannel.MethodCallHandler, SuperPlayerView.OnSuperPlayerViewCallback, ISuperPlayerListener {

    private       SuperPlayerView                    mSuperPlayerView;
    private       FlutterPlugin.FlutterPluginBinding mFlutterPluginBinding;
    private final MethodChannel                      mMethodChannel;
    private final EventChannel                       mEventChannel;
    private final EventChannel mNetChannel;
    private final FTXPlayerEventSink                 mEventSink = new FTXPlayerEventSink();
    final private FTXPlayerEventSink mNetStatusSink = new FTXPlayerEventSink();

    public SuperPlatformPlayerView(Context context, Map<String, Object> params, int viewId, FlutterPlugin.FlutterPluginBinding flutterPluginBinding) {
        super();
        mSuperPlayerView = new SuperPlayerView(context);
        mSuperPlayerView.setPlayerViewCallback(this);
        mSuperPlayerView.setSuperPlayerListener(this);
        mMethodChannel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "cloud.tencent.com/superPlayer/" + viewId);
        mMethodChannel.setMethodCallHandler(this);
        mEventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "cloud.tencent.com/superPlayer/event/" + viewId);
        mEventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object o, EventChannel.EventSink eventSink) {
                mEventSink.setEventSinkProxy(eventSink);
            }

            @Override
            public void onCancel(Object o) {
                mEventSink.setEventSinkProxy(null);
            }
        });
//
//        if (params.containsKey("playerConfig")){
//            Map playerConfig = (Map) params.get("playerConfig");
//            SuperPlayerGlobalConfig prefs = SuperPlayerGlobalConfig.getInstance();
//            prefs.playShiftDomain = (String) playerConfig.get("playShiftDomain");
//            prefs.enableHWAcceleration = (boolean) playerConfig.get("hwAcceleration");
//            prefs.renderMode = (int) playerConfig.get("renderMode");
//        }
//
//        SuperPlayerModel model = new SuperPlayerModel();
//        if (params.containsKey("playerModel")) {
//            Map playerModel = (Map) params.get("playerModel");
//            model.url = (String) playerModel.get("videoURL");
//            model.appId = (int) playerModel.get("appId");
//            model.playDefaultIndex = (int) playerModel.get("defaultPlayIndex");
//            if (playerModel.containsKey("videoId")) {
//                Map videoIdMap = (Map) playerModel.get("videoId");
//                SuperPlayerVideoId videoId = new SuperPlayerVideoId();
//                videoId.fileId = (String) videoIdMap.get("videoId");
//                videoId.pSign = (String) videoIdMap.get("psign");
//                model.videoId = videoId;
//            }
//
//            if (playerModel.containsKey("multiVideoURLs")) {
//                List<SuperPlayerModel.SuperPlayerURL> multiURLs = new ArrayList<SuperPlayerModel.SuperPlayerURL>();
//                List<Map> mapURLs = (List<Map>) playerModel.get("multiVideoURLs");
//                for (Map e:mapURLs) {
//                    SuperPlayerModel.SuperPlayerURL url = new SuperPlayerModel.SuperPlayerURL();
//                    url.qualityName = (String) e.get("title");
//                    url.url = (String) e.get("url");
//                    multiURLs.add(url);
//                }
//                model.multiURLs = multiURLs;
//            }
//        }
//
//        mPlayerModel = model;
//        mSuperPlayerView.playWithModel(model);
        mNetChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "cloud.tencent.com/superPlayer/net/" + viewId);
        mNetChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object o, EventChannel.EventSink eventSink) {
                mNetStatusSink.setEventSinkProxy(eventSink);
            }

            @Override
            public void onCancel(Object o) {
                mNetStatusSink.setEventSinkProxy(null);
            }
        });
    }


    @Override
    public void onStartFullScreenPlay() {
        mEventSink.success(getParams("onStartFullScreenPlay", null));
    }

    @Override
    public void onStopFullScreenPlay() {
        mEventSink.success(getParams("onStopFullScreenPlay", null));
    }

    @Override
    public void onClickFloatCloseBtn() {
        mEventSink.success(getParams("onClickFloatCloseBtn", null));
    }

    @Override
    public void onClickSmallReturnBtn() {
        mEventSink.success(getParams("onSuperPlayerBackAction", null));
    }

    @Override
    public void onStartFloatWindowPlay() {
        mEventSink.success(getParams("onStartFloatWindowPlay", null));
    }

    @Override
    public void onPlaying() {
        mEventSink.success(getParams("onSuperPlayerDidStart", null));
    }

    @Override
    public void onPlayEnd() {
        mEventSink.success(getParams("onSuperPlayerDidEnd", null));
    }

    @Override
    public void onError(int code) {
        mEventSink.success(getParams("onSuperPlayerError", null));
    }

    @Override
    public View getView() {
        return mSuperPlayerView;
    }

    @Override
    public void dispose() {
        mSuperPlayerView.resetPlayer();
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        if (call.method.equals("reloadView")) {
//            String url = call.argument("url");
//            Integer appId = call.argument("appId");
//            String fileId = call.argument("fileId");
//            String psign = call.argument("psign");
//            reloadView(url, appId, fileId, psign);
            result.success(null);
        } else if (call.method.equals("play")) {
            Map playerModel = call.argument("playerModel");
            playWithModel(playerModel);
            result.success(null);
        } else if (call.method.equals("playConfig")) {
            Map playConfig = call.argument("config");
            setPlayConfig(playConfig);
            result.success(null);
        } else if (call.method.equals("disableGesture")) {
            Boolean enable = call.argument("enable");
            disableGesture(enable);
            result.success(null);
        } else if (call.method.equals("setIsAutoPlay")) {
            Boolean enable = call.argument("isAutoPlay");
            setIsAutoPlay(enable);
            result.success(null);
        } else if (call.method.equals("setStartTime")) {
            Double startTime = call.argument("startTime");
            setStartTime(startTime);
            result.success(null);
        } else if (call.method.equals("setLoop")) {
            Boolean enable = call.argument("loop");
            setLoop(enable);
            result.success(null);
        } else if(call.method.equals("resetPlayer")){
            destory();
            result.success(null);
        } else if(call.method.equals("pause")){
            setPause();
            result.success(null);
        } else if(call.method.equals("resume")){
            setResume();
            result.success(null);
        } else if(call.method.equals("stop")){
            setStop();
            result.success(null);
        }else {
            result.notImplemented();
        }
    }

//    public void reloadView(String url, Integer appId, String fileId, String psign) {
//        if (mPlayerModel == null) {
//            return;
//        }
//
//        if (url != null && !url.isEmpty()) {
//            mPlayerModel.videoId = null;
//            mPlayerModel.url = url;
//        }else if (appId > 0 && fileId != null && !fileId.isEmpty()) {
//            SuperPlayerVideoId videoId = new SuperPlayerVideoId();
//            mPlayerModel.appId = (int) appId;
//            videoId.fileId = fileId;
//            if (psign != null && !psign.isEmpty()) {
//                videoId.pSign = psign;
//            }
//            mPlayerModel.videoId = videoId;
//            mPlayerModel.url = null;
//        }
//
//        mSuperPlayerView.resetPlayer();
//        mSuperPlayerView.playWithModel(mPlayerModel);
//    }

    public void playWithModel(Map playerModel) {
        SuperPlayerModel model = new SuperPlayerModel();
        model.url = (String) playerModel.get("videoURL");
        model.appId = (int) playerModel.get("appId");
        if (model.appId > 0) {
            TXLiveBase.setAppID("" + model.appId);
        }
        model.playDefaultIndex = (int) playerModel.get("defaultPlayIndex");
        model.title = (String) playerModel.get("title");
        if (playerModel.containsKey("videoId")) {
            Map videoIdMap = (Map) playerModel.get("videoId");
            SuperPlayerVideoId videoId = new SuperPlayerVideoId();
            videoId.fileId = (String) videoIdMap.get("fileId");
            videoId.pSign = (String) videoIdMap.get("psign");
            model.videoId = videoId;
            model.url = null;
        }

        if (playerModel.containsKey("multiVideoURLs")) {
            List<SuperPlayerModel.SuperPlayerURL> multiURLs = new ArrayList<SuperPlayerModel.SuperPlayerURL>();
            List<Map> mapURLs = (List<Map>) playerModel.get("multiVideoURLs");
            for (Map e : mapURLs) {
                SuperPlayerModel.SuperPlayerURL url = new SuperPlayerModel.SuperPlayerURL();
                url.qualityName = (String) e.get("title");
                url.url = (String) e.get("url");
                multiURLs.add(url);
            }
            model.multiURLs = multiURLs;
        }

        mSuperPlayerView.resetPlayer();
        mSuperPlayerView.playWithModel(model);
    }

    public void setPlayConfig(Map params) {
        SuperPlayerGlobalConfig prefs = SuperPlayerGlobalConfig.getInstance();
        prefs.playShiftDomain = (String) params.get("playShiftDomain");
        prefs.enableHWAcceleration = (boolean) params.get("hwAcceleration");
        prefs.renderMode = (int) params.get("renderMode");
        prefs.mirror = (boolean) params.get("mirror");
        prefs.mute = (boolean) params.get("mute");
        prefs.enableLog = (boolean) params.get("enableLog");

        int maxCacheItem = (int) params.get("maxCacheItem");
        if (maxCacheItem > 0) {
            prefs.maxCacheItem = maxCacheItem;
        }
        Map<String, String> headers = (Map<String, String>) params.get("headers");
        if (headers == null) {
            headers = new HashMap<>();
        }
        prefs.headers = headers;

        double playRate = (double)params.get("playRate");
        if (playRate > 0) {
            prefs.playRate = (float)playRate;
        }
    }

    public void setIsAutoPlay(boolean b) {
        mSuperPlayerView.setIsAutoPlay(b);
    }

    public void setStartTime(double startTime) {
        mSuperPlayerView.setStartTime(startTime);
    }

    /**
     * 关闭所有手势
     *
     * @param flag true为关闭手势，false为开启手势
     */
    public void disableGesture(boolean flag) {
        mSuperPlayerView.disableGesture(flag);
    }

    public void setLoop(boolean b) {
        mSuperPlayerView.setLoop(b);
    }

    public void setPause() {
        mSuperPlayerView.onPause();
    }

    public void setResume() {
        mSuperPlayerView.onResume();
    }

    public void setStop() {
        mSuperPlayerView.resetPlayer();
    }

    private Map<String, Object> getParams(String event, Bundle bundle) {
        Map<String, Object> param = new HashMap();
        if (!event.isEmpty()) {
            param.put("event", event);
        }

        if (bundle != null && !bundle.isEmpty()) {
            Set<String> keySet = bundle.keySet();
            for (String key : keySet) {
                Object val = bundle.get(key);
                param.put(key, val);
            }
        }

        return param;
    }

    private void destory() {
        mSuperPlayerView.resetPlayer();
        mSuperPlayerView = null;
        mMethodChannel.setMethodCallHandler(null);
        mEventChannel.setStreamHandler(null);
    }

    @Override
    public void onVodPlayEvent(TXVodPlayer player, int event, Bundle param) {
        mEventSink.success(getParams("onVodPlayEvent", param));
    }

    @Override
    public void onVodNetStatus(TXVodPlayer player, Bundle status) {
        mNetStatusSink.success(getParams("onVodNetStatus", status));
    }

    @Override
    public void onLivePlayEvent(int event, Bundle param) {
        mEventSink.success(getParams("onLivePlayEvent", param));
    }

    @Override
    public void onLiveNetStatus(Bundle status) {
        mNetStatusSink.success(getParams("onLiveNetStatus", status));
    }
}

File: Flutter/ios/Classes/SuperPlatformPlayerView.m
//
//  SuperPlatformView.m
//  Pods
//
//  Created by Zhirui Ou on 2021/3/10.
//

#import "SuperPlatformPlayerView.h"
#import "SuperPlayer.h"
#import "FTXBasePlayer.h"
#import "FTXPlayerEventSinkQueue.h"

@interface SuperPlatformPlayerView ()<SuperPlayerDelegate, FlutterStreamHandler>
{
    FTXPlayerEventSinkQueue *_eventSink;
}

@property (nonatomic, strong) SuperPlayerView *realPlayerView;
@property (nonatomic, strong) UIView *playerFatherView;
@property (nonatomic, strong) FlutterMethodChannel *methodChannel;
@property (nonatomic, strong) FlutterEventChannel *eventChannel;

@end

@implementation SuperPlatformPlayerView

+ (NSObject<FlutterPlatformView> *)createWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id)args registrar:(id<FlutterPluginRegistrar>)registrar
{
    SuperPlatformPlayerView *playerView = [[SuperPlatformPlayerView alloc] initWithRegistrar:registrar viewIdentifier:viewId];
    /*
    SuperPlayerView *realPlayerView = playerView.realPlayerView;
    if ([args isKindOfClass:[NSDictionary class]]) {
        if ([args[@"playerConfig"] isKindOfClass:[NSDictionary class]]) {
            realPlayerView.playerConfig.playShiftDomain = args[@"playerConfig"][@"playShiftDomain"];
            realPlayerView.playerConfig.mirror = [args[@"playerConfig"][@"mirror"] boolValue];
            realPlayerView.playerConfig.hwAcceleration = [args[@"playerConfig"][@"hwAcceleration"] boolValue];
            realPlayerView.playerConfig.playRate = [args[@"playerConfig"][@"playRate"] floatValue];
            realPlayerView.playerConfig.mute = [args[@"playerConfig"][@"mute"] boolValue];
            realPlayerView.playerConfig.renderMode = [args[@"playerConfig"][@"renderMode"] intValue];
            realPlayerView.playerConfig.headers = args[@"playerConfig"][@"headers"];
            realPlayerView.playerConfig.maxCacheItem = [args[@"playerConfig"][@"maxCacheItem"] intValue];
            realPlayerView.playerConfig.enableLog = [args[@"playerConfig"][@"enableLog"] boolValue];
        }
        
        SuperPlayerModel *model = SuperPlayerModel.new;
        if ([args[@"playerModel"] isKindOfClass:[NSDictionary class]]) {
            model.videoURL = args[@"playerModel"][@"videoURL"];
            model.appId = [args[@"playerModel"][@"appId"] unsignedLongLongValue];
            model.defaultPlayIndex = [args[@"playerModel"][@"defaultPlayIndex"] unsignedLongLongValue];
            if ([args[@"playerModel"][@"videoId"] isKindOfClass:[NSDictionary class]]) {
                SuperPlayerVideoId *videoId = [SuperPlayerVideoId new];
                videoId.fileId = args[@"playerModel"][@"videoId"][@"fileId"];
                videoId.psign = args[@"playerModel"][@"videoId"][@"psign"];
                model.videoId = videoId;
            }
            if ([args[@"playerModel"][@"multiVideoURLs"] isKindOfClass:[NSArray class]]) {
                NSMutableArray *multiVideoURLs = @[].mutableCopy;
                for (NSDictionary *urlDic in args[@"playerModel"][@"multiVideoURLs"]) {
                    SuperPlayerUrl *url = [SuperPlayerUrl new];
                    url.title = urlDic[@"title"];
                    url.url = urlDic[@"url"];
                    [multiVideoURLs addObject:url];
                }
                model.multiVideoURLs = multiVideoURLs;
            }
        }
        
        realPlayerView.autoPlay = [args[@"autoPlay"] boolValue];
        realPlayerView.startTime = [args[@"startTime"] doubleValue];
//        realPlayerView.isLockScreen = [args[@"isLockScreen"] boolValue];
        realPlayerView.isLockScreen = true;
        realPlayerView.disableGesture = [args[@"disableGesture"] boolValue];
        realPlayerView.loop = [args[@"loop"] boolValue];
//        [realPlayerView setFullScreen:[args[@"isFullScreen"] boolValue]];
        realPlayerView.isFullScreen = false;
        
        //realPlayerView.controlView = SuperPlayerControlView.new;
        realPlayerView.repeatBackBtn.hidden = true;
        
        
        [realPlayerView playWithModel:model];
    }
    */

    return playerView;
}
 
- (instancetype)initWithRegistrar:(id<FlutterPluginRegistrar>)registrar viewIdentifier:(int64_t)viewId
{
    if (self = [self init]) {
        __weak typeof(self) weakSelf = self;
        _eventSink = [FTXPlayerEventSinkQueue new];
        _methodChannel = [FlutterMethodChannel methodChannelWithName:[@"cloud.tencent.com/superPlayer/" stringByAppendingString:[NSString stringWithFormat:@"%@", @(viewId)]] binaryMessenger:[registrar messenger]];
        [_methodChannel setMethodCallHandler:^(FlutterMethodCall * _Nonnull call, FlutterResult  _Nonnull result) {
            [weakSelf handleMethodCall:call result:result];
        }];
        _eventChannel = [FlutterEventChannel eventChannelWithName:[@"cloud.tencent.com/superPlayer/event/" stringByAppendingString:[NSString stringWithFormat:@"%@", @(viewId)]] binaryMessenger:[registrar messenger]];
        [_eventChannel setStreamHandler:self];
    }
    
    return self;
}

- (void)dealloc
{
    [self destory];
}

- (void)destory
{
    [_realPlayerView resetPlayer];
    [_eventChannel setStreamHandler:nil];
    _realPlayerView = nil;
    _eventChannel = nil;
    _methodChannel = nil;
    _eventSink = nil;
}

- (void)reloadView:(NSString *)url appId:(long)appId fileId:(NSString *)fileId psign:(NSString *)psign
{
    SuperPlayerModel *playerModel = self.realPlayerView.playerModel;
    if (url.length > 0) {//有url就直接播放
        playerModel.videoURL = url;
        playerModel.videoId = nil;
    }else if (appId > 0 && fileId.length > 0) {
        SuperPlayerVideoId *videoId = [SuperPlayerVideoId new];
        playerModel.appId = appId;
        videoId.fileId = fileId;
        if (psign.length > 0) {
            videoId.psign = psign;
        }
        playerModel.videoId = videoId;
        playerModel.videoURL = nil;
    }

    [self.realPlayerView resetPlayer];
    [self.realPlayerView playWithModel:playerModel];
}

- (void)playWithModel:(NSDictionary *)args
{
    SuperPlayerModel *model = SuperPlayerModel.new;
    if ([args[@"playerModel"] isKindOfClass:[NSDictionary class]]) {
        model.videoURL = args[@"playerModel"][@"videoURL"];
        model.appId = [args[@"playerModel"][@"appId"] longValue];
        model.defaultPlayIndex = [args[@"playerModel"][@"defaultPlayIndex"] unsignedIntValue];
        if ([args[@"playerModel"][@"videoId"] isKindOfClass:[NSDictionary class]]) {
            SuperPlayerVideoId *videoId = [SuperPlayerVideoId new];
            videoId.fileId = args[@"playerModel"][@"videoId"][@"fileId"];
            videoId.psign = args[@"playerModel"][@"videoId"][@"psign"];
            model.videoId = videoId;
            model.videoURL = nil;
        }
        if ([args[@"playerModel"][@"multiVideoURLs"] isKindOfClass:[NSArray class]]) {
            NSMutableArray *multiVideoURLs = @[].mutableCopy;
            for (NSDictionary *urlDic in args[@"playerModel"][@"multiVideoURLs"]) {
                SuperPlayerUrl *url = [SuperPlayerUrl new];
                url.title = urlDic[@"title"];
                url.url = urlDic[@"url"];
                [multiVideoURLs addObject:url];
            }
            model.multiVideoURLs = multiVideoURLs;
        }
    }
    
    [self.realPlayerView resetPlayer];
    [self.realPlayerView playWithModel:model];
}

- (void)setPlayConfig:(NSDictionary *)args
{
    if ([args[@"config"] isKindOfClass:[NSDictionary class]]) {
        self.realPlayerView.playerConfig.playShiftDomain = args[@"config"][@"playShiftDomain"];
        self.realPlayerView.playerConfig.mirror = [args[@"config"][@"mirror"] boolValue];
        self.realPlayerView.playerConfig.hwAcceleration = [args[@"config"][@"hwAcceleration"] boolValue];
        self.realPlayerView.playerConfig.playRate = [args[@"config"][@"playRate"] floatValue];
        self.realPlayerView.playerConfig.mute = [args[@"config"][@"mute"] boolValue];
        self.realPlayerView.playerConfig.renderMode = [args[@"config"][@"renderMode"] intValue];
        self.realPlayerView.playerConfig.headers = args[@"config"][@"headers"];
        self.realPlayerView.playerConfig.maxCacheItem = [args[@"config"][@"maxCacheItem"] intValue];
        self.realPlayerView.playerConfig.enableLog = [args[@"config"][@"enableLog"] boolValue];
    }
}

- (void)setIsAutoPlay:(BOOL)b
{
    self.realPlayerView.autoPlay = b;
}

- (void)setStartTime:(float)startTime
{
    [self.realPlayerView setStartTime:startTime];
}

- (void)disableGesture:(BOOL)b
{
    self.realPlayerView.disableGesture = b;
}

- (void)setLoop:(BOOL)bLoop
{
    self.realPlayerView.loop = bLoop;
}

#pragma mark - FlutterStreamHandler

- (FlutterError* _Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:(FlutterEventSink)events
{
    [_eventSink setDelegate:events];
    return nil;
}

- (FlutterError* _Nullable)onCancelWithArguments:(id _Nullable)arguments
{
    [_eventSink setDelegate:nil];
    return nil;
}

#pragma mark -

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result
{
    NSDictionary *args = call.arguments;
    
    if ([@"reloadView" isEqualToString:call.method]) {
        NSString *url = args[@"url"];
        long appId = [args[@"appId"] longValue];
        NSString *fileId = args[@"fileId"];
        NSString *psign = args[@"psign"];
        [self reloadView:url appId:appId fileId:fileId psign:psign];
        result(nil);
    }else if ([@"play" isEqualToString:call.method]) {
        [self playWithModel:args];
        result(nil);
    }else if ([@"playConfig" isEqualToString:call.method]) {
        [self setPlayConfig:args];
        result(nil);
    }else if ([@"setIsAutoPlay" isEqualToString:call.method]) {
        BOOL isAutoPlay = [args[@"isAutoPlay"] boolValue];
        [self setIsAutoPlay:isAutoPlay];
        result(nil);
    }else if ([@"setStartTime" isEqualToString:call.method]) {
        float startTime = [args[@"startTime"] floatValue];
        [self setStartTime:startTime];
        result(nil);
    }else if ([@"disableGesture" isEqualToString:call.method]) {
        BOOL enable = [args[@"enable"] boolValue];
        [self disableGesture:enable];
        result(nil);
    }else if ([@"setLoop" isEqualToString:call.method]) {
        BOOL loop = [args[@"loop"] boolValue];
        [self setLoop:loop];
        result(nil);
    }else if ([@"resetPlayer" isEqualToString:call.method]) {
        [self destory];
        result(nil);
    }else {
        result(FlutterMethodNotImplemented);
    }
}

#pragma mark - FlutterPlatformView

- (UIView *)view
{
    return self.playerFatherView;
}

#pragma mark - lazy

- (SuperPlayerView *)realPlayerView
{
    if (!_realPlayerView) {
        _realPlayerView = [SuperPlayerView new];
        _realPlayerView.delegate = self;
        _realPlayerView.fatherView = self.playerFatherView;
        [self.playerFatherView addSubview:_realPlayerView];
    }
    
    
    
    return _realPlayerView;
}

- (UIView *)playerFatherView
{
    if (!_playerFatherView) {
        _playerFatherView = UIView.new;
        _playerFatherView.backgroundColor = UIColor.blackColor;
    }
    
    return _playerFatherView;
}

#pragma mark - SuperPlayerDelegate

/// 返回事件
- (void)superPlayerBackAction:(SuperPlayerView *)player
{
    [_eventSink success:[self getParamsWithEvent:@"onSuperPlayerBackAction" withParams:@{}]];
}

/// 全屏改变通知
- (void)superPlayerFullScreenChanged:(SuperPlayerView *)player
{
    if (!player.isFullScreen) {
        [_eventSink success:[self getParamsWithEvent:@"onStopFullScreenPlay" withParams:@{}]];
    }else {
        [_eventSink success:[self getParamsWithEvent:@"onStartFullScreenPlay" withParams:@{}]];
    }
}

/// 播放开始通知
- (void)superPlayerDidStart:(SuperPlayerView *)player;
{
    [_eventSink success:[self getParamsWithEvent:@"onSuperPlayerDidStart" withParams:@{}]];
}

/// 播放结束通知
- (void)superPlayerDidEnd:(SuperPlayerView *)player
{
    [_eventSink success:[self getParamsWithEvent:@"onSuperPlayerDidEnd" withParams:@{}]];
}

/// 播放错误通知
- (void)superPlayerError:(SuperPlayerView *)player errCode:(int)code errMessage:(NSString *)why;
{
    [_eventSink success:[self getParamsWithEvent:@"onSuperPlayerError" withParams:@{
        @"errCode":@(code),
        @"errMessage":why
    }]];
}
// 需要通知到父view的事件在此添加

- (NSDictionary *)getParamsWithEvent:(NSString *)evtName withParams:(NSDictionary *)params
{
    NSMutableDictionary<NSString*,NSObject*> *dict = [NSMutableDictionary dictionaryWithObject:evtName forKey:@"event"];
    if (params != nil && params.count != 0) {
        [dict addEntriesFromDictionary:params];
    }
    return dict;
}


+ (UINavigationController *)currentNavigationController
{
    UINavigationController *currentNav = [self getNearestNavigation:[self appRootViewController]];
    do {
        UINavigationController *subNav = [self getNearestNavigation:[currentNav.viewControllers lastObject]];
        if (subNav) {
            currentNav = subNav;
        } else {
            break;
        }
    } while (1) ;
    
    return currentNav;
}

+ (UIViewController *)topViewController
{
    return [self topViewController:[self appRootViewController]];
}

+ (UINavigationController *)getNearestNavigation:(UIViewController *)rootViewController
{
    if ([rootViewController isKindOfClass:[UINavigationController class]]) {
        return (UINavigationController *)rootViewController;
    }
    if ([rootViewController isKindOfClass:[UITabBarController class]]) {
        UITabBarController *tabController = (UITabBarController *)rootViewController;
        return [self getNearestNavigation:tabController.selectedViewController];
    }
    if (rootViewController.presentedViewController) {
        return [self getNearestNavigation:rootViewController.presentedViewController];
    }
    return nil;
}

+ (UIViewController *)topViewController:(UIViewController *)rootViewController
{
    if ([rootViewController isKindOfClass:[UINavigationController class]]) {
        UINavigationController *navigationController = (UINavigationController *)rootViewController;
        return [self topViewController:[navigationController.viewControllers lastObject]];
    }
    if ([rootViewController isKindOfClass:[UITabBarController class]]) {
        UITabBarController *tabController = (UITabBarController *)rootViewController;
        return [self topViewController:tabController.selectedViewController];
    }
    if (rootViewController.presentedViewController) {
        return [self topViewController:rootViewController.presentedViewController];
    }
    return rootViewController;
}


+ (UIViewController *)appRootViewController
{
    UIViewController *root = [UIApplication sharedApplication].keyWindow.rootViewController;
    if (!root) {
        root = [UIApplication sharedApplication].delegate.window.rootViewController;
    }
    return root;
}

@end
